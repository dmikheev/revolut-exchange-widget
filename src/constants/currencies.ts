export enum Currency {
    USD = 'USD',
    EUR = 'EUR',
    GBP = 'GBP',
}

export const SymbolByCurrency = {
    [Currency.USD]: '$',
    [Currency.EUR]: '€',
    [Currency.GBP]: '£',
};
