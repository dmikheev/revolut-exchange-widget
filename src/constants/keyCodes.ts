export enum KeyCode {
  ENTER = 13,
  DIGIT_1 = 49,
}
