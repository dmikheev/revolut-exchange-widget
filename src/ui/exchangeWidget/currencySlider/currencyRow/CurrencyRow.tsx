import {
  createStyles,
  TextField,
  Theme,
  Typography,
  WithStyles,
  withStyles,
} from '@material-ui/core';
import { TextFieldProps } from '@material-ui/core/TextField';
import classNames from 'classnames';
import React from 'react';
import { Currency } from '../../../../constants/currencies';

import styles from './CurrencyRow.module.css';

export interface ICurrencyRowProps {
  className?: string;
  balance: number;
  currency: Currency;
  isRateDisabled?: boolean;
}
const CurrencyRow: React.FC<ICurrencyRowProps> = ({
  className,
  balance,
  currency,
}) => {
  return (
    <div className={classNames(styles.wrap, className)}>
      <div className={styles.column}>
        <Typography variant="h3">
          {currency}
        </Typography>
        <Typography variant="body2" className={styles.note}>
          You have {balance} {currency}
        </Typography>
      </div>
      <div className={styles.column}>
        <StyledTextField/>
      </div>
    </div>
  );
};
export default CurrencyRow;

const textFieldStyles = (theme: Theme) => createStyles({
  input: theme.typography.h3,
});

type IStyledTextFieldComponentProps = TextFieldProps & WithStyles<typeof textFieldStyles>;
const StyledTextFieldComponent: React.FC<IStyledTextFieldComponentProps> =
  ({ classes, ...restProps }) => {
    const inputProps = {
      classes: {
        root: classes.input,
      },
    };

    return (
      <TextField {...restProps} InputProps={inputProps}/>
    );
  };
const StyledTextField = withStyles(textFieldStyles)(StyledTextFieldComponent);
