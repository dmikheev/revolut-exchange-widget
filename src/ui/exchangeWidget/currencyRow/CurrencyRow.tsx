import {
  CircularProgress,
  createStyles,
  FormControl,
  InputLabel,
  MenuItem,
  Select,
  Theme,
  Typography,
  WithStyles,
  withStyles,
} from '@material-ui/core';
import { TextFieldProps } from '@material-ui/core/TextField';
import classNames from 'classnames';
import React from 'react';
import { Currency, SymbolByCurrency } from '../../../constants/currencies';
import { KeyCode } from '../../../constants/keyCodes';
import { cashFormat } from '../../../utils/cashFormat';
import CashInput from './CashInput';

import styles from './CurrencyRow.module.css';

interface ICurrencyRowProps {
  className?: string;
  amountStr: string;
  balance: number;
  currencies: Currency[];
  currency: Currency;
  sourceCurrency: Currency;
  isBalanceError?: boolean;
  isRateFetching?: boolean;
  isSourceCurrency?: boolean;
  rate?: number;
  onAmountChange(amount: string): void;
  onCurrencyChange(currency: Currency): void;
  onInputEnterPress?(): void;
}

const CurrencyRow: React.FC<ICurrencyRowProps> = ({
  className,
  amountStr,
  balance,
  currencies,
  currency,
  sourceCurrency,
  isBalanceError,
  isRateFetching,
  isSourceCurrency,
  rate,
  onAmountChange,
  onCurrencyChange,
  onInputEnterPress,
}) => {
  const onCurrencySelectChange = React.useCallback(
    (event: React.ChangeEvent<HTMLSelectElement>) => onCurrencyChange(event.target.value as Currency),
    [onCurrencyChange],
  );
  const onInputChange = React.useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => onAmountChange(event.target.value),
    [onAmountChange],
  );
  const onInputKeyPress = React.useCallback(
    (event: React.KeyboardEvent<HTMLInputElement>) => {
      if (onInputEnterPress && event.which === KeyCode.ENTER) {
        onInputEnterPress();
      }
    },
    [onInputEnterPress],
  );

  const selectItemsHtml = currencies.map((cur) => (
    <MenuItem key={cur} value={cur}>{cur}</MenuItem>
  ));

  let amountColumnHtml;
  if (isSourceCurrency || currency !== sourceCurrency) {
    let rateContentHtml;
    if (!isSourceCurrency) {
      const rateValueHtml = rate !== undefined && (
        <Typography variant="body2" className={styles.note}>
          {SymbolByCurrency[currency]}1 = {SymbolByCurrency[sourceCurrency]}{cashFormat(rate)}
        </Typography>
      );
      const rateProgressHtml = isRateFetching && (
        <CircularProgress className={styles.rate_progress} size="0.8em" color="secondary"/>
      );

      rateContentHtml = (
        <React.Fragment>
          {rateValueHtml}
          {rateProgressHtml}
        </React.Fragment>
      );
    }

    amountColumnHtml = (
      <React.Fragment>
        <StyledTextField
          error={isBalanceError}
          value={amountStr}
          InputProps={{inputProps: {className: styles.input}}}
          onChange={onInputChange}
          onKeyPress={onInputKeyPress}
        />
        <div className={styles.rate_wrap}>
          {rateContentHtml}
        </div>
      </React.Fragment>
    );
  }

  return (
    <div className={classNames(styles.wrap, className)}>
      <div className={styles.column}>
        <FormControl>
          <InputLabel>{isSourceCurrency ? 'From' : 'To'}</InputLabel>
          <Select value={currency} onChange={onCurrencySelectChange}>
            {selectItemsHtml}
          </Select>
        </FormControl>
        <Typography variant="body2" className={styles.note} color={isBalanceError ? 'error' : undefined}>
          You have {SymbolByCurrency[currency]}{cashFormat(balance)}
        </Typography>
      </div>
      <div className={styles.column}>
        {amountColumnHtml}
      </div>
    </div>
  );
};
export default CurrencyRow;

const textFieldStyles = (theme: Theme) => createStyles({
  input: theme.typography.h3,
});

type IStyledTextFieldComponentProps = TextFieldProps & WithStyles<typeof textFieldStyles>;
const StyledTextFieldComponent: React.FC<IStyledTextFieldComponentProps> =
  ({classes, ...restProps}) => {
    const inputProps = {
      ...restProps.InputProps,
      classes: {
        root: classes.input,
      },
    };

    return (
      <CashInput {...restProps} InputProps={inputProps}/>
    );
  };
const StyledTextField = withStyles(textFieldStyles)(StyledTextFieldComponent);
