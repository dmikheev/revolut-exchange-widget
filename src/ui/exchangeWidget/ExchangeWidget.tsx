import { Button } from '@material-ui/core';
import classNames from 'classnames';
import React from 'react';
import { Currency } from '../../constants/currencies';
import CurrencyRow from './currencyRow/CurrencyRow';

import styles from './ExchangeWidget.module.css';

interface IExchangeWidgetProps {
  backgroundColor: string;
  className?: string;
  amountFromStr: string;
  amountToStr: string;
  balanceFrom: number;
  balanceTo: number;
  currencies: Currency[];
  currencyFrom: Currency;
  currencyTo: Currency;
  isExchangeDisabled: boolean;
  isRateFetching: boolean;
  isSourceBalanceError: boolean;
  rateTo?: number;

  onAmountFromChange(value: string): void;
  onAmountToChange(value: string): void;
  onCurrencyFromChange(currency: Currency): void;
  onCurrencyToChange(currency: Currency): void;
  onExchange(): void;
}
const ExchangeWidget: React.FC<IExchangeWidgetProps> = ({
  backgroundColor,
  className,
  amountFromStr,
  amountToStr,
  balanceFrom,
  balanceTo,
  currencies,
  currencyFrom,
  currencyTo,
  isExchangeDisabled,
  isRateFetching,
  isSourceBalanceError,
  rateTo,
  onAmountFromChange,
  onAmountToChange,
  onCurrencyFromChange,
  onCurrencyToChange,
  onExchange,
}) => (
  <div className={classNames(styles.wrap, className)} style={{ backgroundColor }}>
    <CurrencyRow
      className={styles.row}
      amountStr={amountFromStr}
      balance={balanceFrom}
      currencies={currencies}
      currency={currencyFrom}
      sourceCurrency={currencyFrom}
      isBalanceError={isSourceBalanceError}
      isSourceCurrency={true}
      onAmountChange={onAmountFromChange}
      onCurrencyChange={onCurrencyFromChange}
      onInputEnterPress={isExchangeDisabled ? undefined : onExchange}
    />
    <CurrencyRow
      className={styles.row}
      amountStr={amountToStr}
      balance={balanceTo}
      currencies={currencies}
      currency={currencyTo}
      sourceCurrency={currencyFrom}
      isRateFetching={isRateFetching}
      isSourceCurrency={false}
      rate={rateTo}
      onAmountChange={onAmountToChange}
      onCurrencyChange={onCurrencyToChange}
      onInputEnterPress={isExchangeDisabled ? undefined : onExchange}
    />
    <div className={styles.button_row}>
      <Button variant="contained" color="primary" disabled={isExchangeDisabled} size="large" onClick={onExchange}>
        Exchange
      </Button>
    </div>
  </div>
);
export default ExchangeWidget;
