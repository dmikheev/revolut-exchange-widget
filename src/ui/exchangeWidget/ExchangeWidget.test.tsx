import { Button } from '@material-ui/core';
import toJson from 'enzyme-to-json';
import React, { MouseEvent, MouseEventHandler } from 'react';
import { shallow } from 'enzyme';
import { Currency } from '../../constants/currencies';
import CurrencyRow from './currencyRow/CurrencyRow';
import ExchangeWidget from './ExchangeWidget';

const currencies = [Currency.USD, Currency.EUR, Currency.GBP];
const nop = () => {};
const testProps = {
  backgroundColor: '#fff',
  amountFromStr: '111',
  balanceFrom: 222,
  amountToStr: '333',
  balanceTo: 444,
  currencies: currencies,
  currencyFrom: Currency.GBP,
  currencyTo: Currency.USD,
  isExchangeDisabled: false,
  isRateFetching: false,
  isSourceBalanceError: false,
  rateTo: 0.75,
  onAmountFromChange: nop,
  onAmountToChange: nop,
  onCurrencyFromChange: nop,
  onCurrencyToChange: nop,
  onExchange: nop,
};

it('renders correctly', () => {
  const wrapper = shallow(<ExchangeWidget {...testProps}/>);
  expect(toJson(wrapper)).toMatchSnapshot();
});

it('calls onAmountFromChange when the amount in the first row changes', () => {
  const amountStr = '55';
  const onAmountFromChange = jest.fn();
  const wrapper = shallow(<ExchangeWidget {...testProps} onAmountFromChange={onAmountFromChange}/>);

  const rows = wrapper.find(CurrencyRow);
  expect(rows).toHaveLength(2);

  rows.at(0).props().onAmountChange(amountStr);

  expect(onAmountFromChange).toBeCalledTimes(1);
  expect(onAmountFromChange).toHaveBeenLastCalledWith(amountStr);
});

it('calls onAmountToChange when the amount in the second row changes', () => {
  const amountStr = '66';
  const onAmountToChange = jest.fn();
  const wrapper = shallow(<ExchangeWidget {...testProps} onAmountToChange={onAmountToChange}/>);

  const rows = wrapper.find(CurrencyRow);
  expect(rows).toHaveLength(2);

  rows.at(1).props().onAmountChange(amountStr);

  expect(onAmountToChange).toBeCalledTimes(1);
  expect(onAmountToChange).toHaveBeenLastCalledWith(amountStr);
});

it('calls onCurrencyFromChange when the currency in the first row changes', () => {
  const currency = Currency.EUR;
  const onCurrencyFromChange = jest.fn();
  const wrapper = shallow(<ExchangeWidget {...testProps} onCurrencyFromChange={onCurrencyFromChange}/>);

  const rows = wrapper.find(CurrencyRow);
  expect(rows).toHaveLength(2);

  rows.at(0).props().onCurrencyChange(currency);

  expect(onCurrencyFromChange).toBeCalledTimes(1);
  expect(onCurrencyFromChange).toHaveBeenLastCalledWith(currency);
});

it('calls onCurrencyToChange when the currency in the second row changes', () => {
  const currency = Currency.GBP;
  const onCurrencyToChange = jest.fn();
  const wrapper = shallow(<ExchangeWidget {...testProps} onCurrencyToChange={onCurrencyToChange}/>);

  const rows = wrapper.find(CurrencyRow);
  expect(rows).toHaveLength(2);

  rows.at(1).props().onCurrencyChange(currency);

  expect(onCurrencyToChange).toBeCalledTimes(1);
  expect(onCurrencyToChange).toHaveBeenLastCalledWith(currency);
});

it('calls onExchange when the button is clicked', () => {
  const onExchange = jest.fn();
  const wrapper = shallow(<ExchangeWidget {...testProps} isExchangeDisabled={false} onExchange={onExchange}/>);

  const button = wrapper.find(Button);
  expect(button).toHaveLength(1);

  expect(button.props().onClick).toBeTruthy();
  button.props().onClick!({} as MouseEvent<HTMLButtonElement>);

  expect(onExchange).toBeCalledTimes(1);
});

it('calls onExchange when Enter is pressed in one of the inputs', () => {
  const onExchange = jest.fn();
  const wrapper = shallow(<ExchangeWidget {...testProps} isExchangeDisabled={false} onExchange={onExchange}/>);

  const rows = wrapper.find(CurrencyRow);
  expect(rows).toHaveLength(2);

  expect(rows.at(0).props().onInputEnterPress).toBeTruthy();
  rows.at(0).props().onInputEnterPress!();
  expect(onExchange).toBeCalledTimes(1);

  onExchange.mockClear();

  expect(rows.at(1).props().onInputEnterPress).toBeTruthy();
  rows.at(1).props().onInputEnterPress!();
  expect(onExchange).toBeCalledTimes(1);
});

it('disables the button according to isExchangeDisabled prop', () => {
  const isExchangeButtonDisabled1 = true;
  const wrapper1 = shallow(<ExchangeWidget {...testProps} isExchangeDisabled={isExchangeButtonDisabled1}/>);

  const button1 = wrapper1.find(Button);
  expect(button1).toHaveLength(1);
  expect(button1.props().disabled).toEqual(isExchangeButtonDisabled1);

  const isExchangeButtonDisabled2 = false;
  const wrapper2 = shallow(<ExchangeWidget {...testProps} isExchangeDisabled={isExchangeButtonDisabled2}/>);

  const button2 = wrapper2.find(Button);
  expect(button2).toHaveLength(1);
  expect(button2.props().disabled).toEqual(isExchangeButtonDisabled2);
});

it('doesn\'t forward onInputEnterPress when isExchangeDisabled === true', () => {
  const wrapper = shallow(<ExchangeWidget {...testProps} isExchangeDisabled={true}/>);

  const rows = wrapper.find(CurrencyRow);
  expect(rows).toHaveLength(2);

  expect(rows.at(0).props().onInputEnterPress).toBeFalsy();
  expect(rows.at(1).props().onInputEnterPress).toBeFalsy();
});

it('forwards the isRateFetching prop', () => {
  const isRateFetching1 = true;
  const wrapper1 = shallow(<ExchangeWidget {...testProps} isRateFetching={isRateFetching1}/>);

  const rows1 = wrapper1.find(CurrencyRow);
  expect(rows1).toHaveLength(2);
  expect(rows1.at(1).props().isRateFetching).toEqual(isRateFetching1);

  const isRateFetching2 = false;
  const wrapper2 = shallow(<ExchangeWidget {...testProps} isRateFetching={isRateFetching2}/>);

  const rows2 = wrapper2.find(CurrencyRow);
  expect(rows2).toHaveLength(2);
  expect(rows2.at(1).props().isRateFetching).toEqual(isRateFetching2);
});

it('forwards the isSourceBalanceError prop', () => {
  const isSourceBalanceError1 = true;
  const wrapper1 = shallow(<ExchangeWidget {...testProps} isSourceBalanceError={isSourceBalanceError1}/>);

  const rows1 = wrapper1.find(CurrencyRow);
  expect(rows1).toHaveLength(2);
  expect(rows1.at(0).props().isBalanceError).toEqual(isSourceBalanceError1);

  const isSourceBalanceError2 = false;
  const wrapper2 = shallow(<ExchangeWidget {...testProps} isSourceBalanceError={isSourceBalanceError2}/>);

  const rows2 = wrapper2.find(CurrencyRow);
  expect(rows2).toHaveLength(2);
  expect(rows2.at(0).props().isBalanceError).toEqual(isSourceBalanceError2);
});
