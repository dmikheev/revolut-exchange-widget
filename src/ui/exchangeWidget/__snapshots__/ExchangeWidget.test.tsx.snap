// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="wrap"
  style={
    Object {
      "backgroundColor": "#fff",
    }
  }
>
  <CurrencyRow
    amountStr="111"
    balance={222}
    className="row"
    currencies={
      Array [
        "USD",
        "EUR",
        "GBP",
      ]
    }
    currency="GBP"
    isBalanceError={false}
    isSourceCurrency={true}
    onAmountChange={[Function]}
    onCurrencyChange={[Function]}
    onInputEnterPress={[Function]}
    sourceCurrency="GBP"
  />
  <CurrencyRow
    amountStr="333"
    balance={444}
    className="row"
    currencies={
      Array [
        "USD",
        "EUR",
        "GBP",
      ]
    }
    currency="USD"
    isRateFetching={false}
    isSourceCurrency={false}
    onAmountChange={[Function]}
    onCurrencyChange={[Function]}
    onInputEnterPress={[Function]}
    rate={0.75}
    sourceCurrency="GBP"
  />
  <div
    className="button_row"
  >
    <WithStyles(Button)
      color="primary"
      disabled={false}
      onClick={[Function]}
      size="large"
      variant="contained"
    >
      Exchange
    </WithStyles(Button)>
  </div>
</div>
`;
